contracts/upgradeable_contracts/BaseMediatorFeeManager.sol

```diff
@@ -1,164 +1,108 @@
-    address internal constant F_ADDR = 0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF;
     uint256 internal constant MAX_REWARD_ACCOUNTS = 50;
 
+    uint256 public fee;
+    address[] internal rewardAccounts;
+
     modifier validFee(uint256 _fee) {
         require(_fee < MAX_FEE);
         /* solcov ignore next */
         _;
     }
 
     /**
-    * @dev Initialize the list of accounts that receives rewards for the mediator operations.
-    * The list of accounts is stored as a Linked List where the list starts and ends at F_ADDR.
-    * Example: F_ADDR -> account1; account1 -> account2; account2 -> F_ADDR
-    * @param _accounts list of accounts
-    */
-    function initializeRewardAccounts(address[] _accounts) public {
-        require(_accounts.length > 0);
-        for (uint256 i = 0; i < _accounts.length; i++) {
-            require(_accounts[i] != address(0) && _accounts[i] != F_ADDR);
-            require(!isRewardAccount(_accounts[i]));
-
-            if (i == 0) {
-                setNextRewardAccount(F_ADDR, _accounts[i]);
-                if (_accounts.length == 1) {
-                    setNextRewardAccount(_accounts[i], F_ADDR);
-                }
-            } else if (i == _accounts.length - 1) {
-                setNextRewardAccount(_accounts[i - 1], _accounts[i]);
-                setNextRewardAccount(_accounts[i], F_ADDR);
-            } else {
-                setNextRewardAccount(_accounts[i - 1], _accounts[i]);
-            }
-        }
-
-        setRewardAccountsCount(_accounts.length);
-    }
-
-    /**
-    * @dev Tells the list of accounts that receives rewards for the operations.
-    * @return the list of reward accounts
+    * @dev Stores the initial parameters of the fee manager.
+    * @param _owner address of the owner of the fee manager contract.
+    * @param _fee the fee percentage amount.
+    * @param _rewardAccountList list of addresses that will receive the fee rewards.
     */
-    function rewardAccounts() external view returns (address[] memory) {
-        address[] memory list = new address[](rewardAccountsCount());
-        uint256 counter = 0;
-        address nextRewardAccount = getNextRewardAccount(F_ADDR);
-        require(nextRewardAccount != address(0));
-
-        while (nextRewardAccount != F_ADDR) {
-            list[counter] = nextRewardAccount;
-            nextRewardAccount = getNextRewardAccount(nextRewardAccount);
-            counter++;
-
-            require(nextRewardAccount != address(0));
-        }
-
-        return list;
+    constructor(address _owner, uint256 _fee, address[] _rewardAccountList) public {
+        require(_rewardAccountList.length > 0 && _rewardAccountList.length <= MAX_REWARD_ACCOUNTS);
+        _transferOwnership(_owner);
+        _setFee(_fee);
+        rewardAccounts = _rewardAccountList;
```

Comment:
What will happen if one of the accounts is the mediator address?

Reply:
Great question!

I performed some tests. Added the home mediator address plus one regular account as reward addresses in home and repeating the same in foreign. Fee amount 0.1%

Test 1, Home to Foreign transfer:
- Transferred 10 native tokens to home mediator.
- In foreign 9.99 tokens were minted to the user and 0.01 to the fee manager
- onTokenTransfer method of fee manager is invoked and the fee manager distributes the fee between the accounts, it transfers the fee to the mediator and calls onTokenTransfer of the mediator. This method tries to bridge the token to home side but in this test, the amount was smaller than minPerTx(), so it fails, and because of this [check](https://github.com/poanetwork/tokenbridge-contracts/blob/443ad4c3db5f2f38d5f9e8437812a314b8454243/contracts/ERC677BridgeToken.sol#L62) the token transfer of the fee is reverted and so the whole `onTokenTransfer` method invocation to the fee manager reverts. So the token fee amount remains locked in the fee manager contract.

Explorer [link](https://kovan.etherscan.io/tx/0xa17ca781613542b741bd8f48f7f4054e7a6d6c047ec5ef858bd28ee0d23d6415
) to the transaction of test 1
[Link](https://dashboard.tenderly.dev/tx/kovan/0xa17ca781613542b741bd8f48f7f4054e7a6d6c047ec5ef858bd28ee0d23d6415) to debug the transaction of test 1 in tenderly

Test 2. This time minPerTx() is set to `1 wei`. Home to Foreign transfer:
- Transferred `10` native tokens to home mediator.
- In foreign `9.99` tokens were minted to the user and 0.01 to the fee manager
- onTokenTransfer method of fee manager is invoked and the fee manager distributes the fee between the accounts, it transfers the fee to the mediator and calls onTokenTransfer of the mediator. This time the amount is within the limits so it is able to send the message to the amb bridge and burn the tokens. The message specifies as the recipient the fee manager address from foreign side and the amount of `0.005`

Explorer [link](https://kovan.etherscan.io/tx/0xe707c0a96c00860ca36d63a4982ab270d182f5d33644d189b0ea611b3a6ace7a
) to the transaction of test 2
[Link](https://dashboard.tenderly.dev/tx/kovan/0xe707c0a96c00860ca36d63a4982ab270d182f5d33644d189b0ea611b3a6ace7a) to debug the transaction of test 2 in tenderly

Test 3. Continuation of test 2
- From test 2, `0.005` tokens were transferred to home with the fee manager address from foreign as the recipient.
- In home, `0.004995` are transferred to the foreign fee manager address (which no one has access to in home)
- `0,000005` is transferred to the home fee manager.
- onTokenTransfer method of fee manager is invoked and the fee manager distributes the fee between the accounts. The fee manager tries to send `0.0000025` native tokens to the mediator but the `Send` fails (I think because of gas required by the fallback method) and so the `Sacrifice` contract was deployed and self-destructed which allowed allocating the fee amount in the mediator contract but it does not invoke any method.
- Now, that `0.0000025` is locked in the mediator contract but no ERC677 token representation will be minted for that amount.
Home Mediator locked balance: `9.9950025`
Total Supply ERC677 in foreign: `9.995`
Explorer [link](https://blockscout.com/poa/sokol/tx/0xffda1763b5c1cd4a5e574759eb3e82eabdff199193855968189e6ca5589c53c8
) to the transaction of test 3

Reply:
Well... What does it mean? Should we avoid situations when the mediator is the reward receiver? Which properties should not the reward receiver have to get the reward without issues?

Reply:
> What does it mean?

To summarize:
If home mediator is used as reward account in home, there will be native tokens locked in the mediator that won't be able to be used.

If foreign mediator is used as reward account in foreign:
- Fee distribution can fail if the fee amount is not within limits (maybe too small or daily limit reached) and the fee will be locked in the fee manager and won't be able to be used
- If amount is within limits, it will send the tokens back to home side. The native tokens in home will be transferred to an account (address of fee manager in foreign) that can't be used. 

>Which properties should not the reward receiver have to get the reward without issues?

To get rewards without issues:
- In the case of home native tokens, it will receive the rewards anyways but ideally, to save gas, the contract should have an empty fallback method so it does not exceed the gas usage provided in the `send` method.
- In the case of foreign erc677 tokens, the only requirement applies to the mediator contract as reward and it is that the `onTokenTransfer` method should not revert. Any other contract that is not the mediator contract can safely not have the `onTokenTransfer` or can revert without affecting the token transfer.

> Should we avoid situations when the mediator is the reward receiver?

So I think the best option is to avoid using the mediator contracts as reward accounts.
Should we add special checks for this in the contract?

One option to send back to home some amount of fee tokens after a Home-to-foreign transfer, if we want to support that case, could be to accumulate the fees in a separate reward receiver contract that will have an `onTokenTransfer` method and once the amount is within limits it can call `transferAndCall(mediatorAddress, accumulatedFees, validRecipientInHome)`

Reply:
Thanks for very good tests you did. They reveal few issues which we need to address:

1. Since we have one-one correspondence between the Fee Manager and the Mediator contract, it makes sense to protect the system to not allow to add the mediator contract address as the reward receiver.

2. Until we have support of multiple transfer requests (https://github.com/poanetwork/tokenbridge-contracts/issues/307) in one transaction let's try to block second, third etc requests on the mediator contract. I would imagine the case when few contracts which are registered as the reward accounts, try to pass the message through the bridge.

3. We need to develop a recovery mechanism to be able to solve the bridge imbalance appearing in the cases similar to described in the test 3. Any contract could use `selfdestruct`-approach so the mediator fallback will not be called. In this case even if the contract on the Home side does not exist any more it is still possible that the corresponding contract (created with `CREATE2` EVM-operation) on the Foreign side is still exist. I would suggest a method that can be called only by the upgradability admin, this method will compare the balance of the bridge mediator contract with the value stored in the contract (we need to introduce such value) and if they differ, allow to send request to another side of the bridge to mint a number of tokens requested in the method. Do you think it will be secure?

Reply:
>Do you think it will be secure?

It looks good to me. One other possible idea could be to extend `claimTokens` method of home side, so instead of sending a request to mint the difference on the other side, send the difference of native tokens to an account provided by parameter in the method

Reply:
> instead of sending a request to mint the difference on the other side, send the difference of native tokens to an account provided by parameter in the method

Yes, from security point of view it is identical actions.